test/core/event_engine/cf/cf_engine_test.cc

```diff
@@ -80,6 +82,211 @@ TEST(CFEventEngineTest, TestConnectionCancelled) {
+
+  EXPECT_TRUE(cf_engine->GetDNSResolver({}).status().ok());
+  EXPECT_TRUE(cf_engine->GetDNSResolver({.dns_server = ""}).status().ok());
+  EXPECT_EQ(
+      cf_engine->GetDNSResolver({.dns_server = "8.8.8.8"}).status().code(),
+      absl::StatusCode::kInvalidArgument);
+  EXPECT_EQ(
+      cf_engine->GetDNSResolver({.dns_server = "8.8.8.8:53"}).status().code(),
+      absl::StatusCode::kInvalidArgument);
+  EXPECT_EQ(
+      cf_engine->GetDNSResolver({.dns_server = "invalid"}).status().code(),
+      absl::StatusCode::kInvalidArgument);
+}
+
+TEST(CFEventEngineTest, TestResolveLocalhost) {
+  grpc_core::Notification resolve_signal;
+
+  auto cf_engine = std::make_shared<CFEventEngine>();
+  auto dns_resolver = cf_engine->GetDNSResolver({});
+
+  dns_resolver.value()->LookupHostname(
+      [&resolve_signal](auto result) {
+        EXPECT_TRUE(result.status().ok());
+        EXPECT_THAT(ResolvedAddressesToStrings(result.value()),
+                    testing::UnorderedElementsAre("127.0.0.1:80", "[::1]:80"));
+
+        resolve_signal.Notify();
+      },
+      "localhost", "80");
+
+  resolve_signal.WaitForNotification();
+}
+
+TEST(CFEventEngineTest, TestResolveRemote) {
+  grpc_core::Notification resolve_signal;
+
+  auto cf_engine = std::make_shared<CFEventEngine>();
+  auto dns_resolver = cf_engine->GetDNSResolver({});
+
+  dns_resolver.value()->LookupHostname(
+      [&resolve_signal](auto result) {
+        EXPECT_TRUE(result.status().ok());
+        EXPECT_THAT(ResolvedAddressesToStrings(result.value()),
+                    testing::UnorderedElementsAre("127.0.0.1:80", "[::1]:80"));
+
+        resolve_signal.Notify();
+      },
+      "localtest.me:80", "443");
+
+  resolve_signal.WaitForNotification();
+}
+
+TEST(CFEventEngineTest, TestResolveIPv4Remote) {
+  grpc_core::Notification resolve_signal;
+
+  auto cf_engine = std::make_shared<CFEventEngine>();
+  auto dns_resolver = cf_engine->GetDNSResolver({});
+
+  dns_resolver.value()->LookupHostname(
+      [&resolve_signal](auto result) {
+        EXPECT_TRUE(result.status().ok());
+        EXPECT_THAT(ResolvedAddressesToStrings(result.value()),
+                    testing::IsSubsetOf(
+                        {"1.2.3.4:80", "[64:ff9b::102:304]:80" /*NAT64*/}));
+
+        resolve_signal.Notify();
+      },
+      "1.2.3.4.nip.io:80", "");
+
+  resolve_signal.WaitForNotification();
+}
+
+TEST(CFEventEngineTest, TestResolveIPv6Remote) {
+  grpc_core::Notification resolve_signal;
+
+  auto cf_engine = std::make_shared<CFEventEngine>();
+  auto dns_resolver = cf_engine->GetDNSResolver({});
+
+  dns_resolver.value()->LookupHostname(
+      [&resolve_signal](auto result) {
+        EXPECT_TRUE(result.status().ok());
+        EXPECT_THAT(
+            ResolvedAddressesToStrings(result.value()),
+            testing::UnorderedElementsAre("[2607:f8b0:400a:801::1002]:80"));
+
+        resolve_signal.Notify();
+      },
+      "2607-f8b0-400a-801--1002.sslip.io.", "80");
```

Comment:
Same as above. Word of caution that any test that relies on an external service may A) become flaky even though the implementation is fine, and B) may be pretty unfriendly if we run `bazel test --runs_per_test=1000 `

True below as well, I won't comment on it.